### PR TITLE
Mixed Type  Case Lists and Case Search

### PR DESCRIPTION
**Pull request by sravfeyn**
## Summary
[Design doc](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=VhlHNC5xOgbQcTwgqwmGDw&ved=0ahUKEwjIpqaU4ozxAhXOgKwCHSy5DQ44ABABKAAwAA&url=https://drive.google.com/a/dimagi.com/open?id%3D1-r6VCViy-cAwlpLIDN_Er7RGcWB0JKb5o821OI74qTM%26usp%3Dchrome_omnibox&usg=AOvVaw1mSc8FlhaiBdtCAdHj9YCF). Ticket https://dimagi-dev.atlassian.net/browse/USH-1014

This adds an option to select multiple case-types to be displayed under Case List and Case Search screens. When the option is enabled cases of multiple types will be displayed in the case-list and case search results.

Screenshot 2021-06-15 at 3 39 52 PM

## Feature Flag
USH_CASE_CLAIM_UPDATES

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-3135

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/app_manager/suite_xml/sections/remote_requests.py [hunk 1]
```diff
@@ -131,8 +131,13 @@ def _build_remote_request_datums(self):
...
         nodeset = CaseTypeXpath(self.module.case_type).case(instance_name=RESULTS_INSTANCE)
-        if self.module.search_config.search_filter and toggles.USH_CASE_CLAIM_UPDATES.enabled(self.app.domain):
-            nodeset = f"{nodeset}[{interpolate_xpath(self.module.search_config.search_filter)}]"
+        if toggles.USH_CASE_CLAIM_UPDATES.enabled(self.app.domain):
+            additional_types = list(set(self.module.additional_case_types) - {self.module.case_type})
+            if additional_types:
+                nodeset = CaseTypeXpath(self.module.case_type).cases(
+                    additional_types, instance_name=RESULTS_INSTANCE)
+            if self.module.search_config.search_filter:
+                nodeset = f"{nodeset}[{interpolate_xpath(self.module.search_config.search_filter)}]"
```
Comment (orangejenny) [inline corehq/apps/app_manager/suite_xml/sections/remote_requests.py:140]: Should this use a `+=`? It looks like if there are additional types and also a search filter, the final nodeset won't incorporate both.

File: corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html [hunk 2]
```diff
@@ -51,6 +51,29 @@ <h4 class="panel-title panel-title-nolink">{% trans "Menu" %}</h4>
...
 
+            {% if request|toggle_enabled:'USH_CASE_CLAIM_UPDATES' %}
+            <div class="form-group">
+              <label class="{% css_label_class %} control-label">
+                 {% trans "Case List and Case Search Types" %}
+                <span class="hq-help-template"
+                      data-title="{% trans "Case List and Case Search Types" %}"
+                      data-content="{% blocktrans %}
+                             Cases of these types will be displayed in Case List and Case Search screens.
+                             <a href='https://confluence.dimagi.com/display/ccinternal/Case+Search+and+Claim'>(More Information)</span>
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:62]:
```suggestion
                             <a target='_blank' href='https://confluence.dimagi.com/display/ccinternal/Case+Search+and+Claim'>(More Information)</a>
```

File: corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html [hunk 3]
```diff
@@ -51,6 +51,29 @@ <h4 class="panel-title panel-title-nolink">{% trans "Menu" %}</h4>
...
+                             {% endblocktrans %}">
+                </span>
+              </label>
+              <div class="{% css_field_class %}">
+                <select name='additional_case_types' class='form-control multiselect-caselist' multiple="multiple" >
+                  {% for case_type in case_types %}
+                    <option value="{{ case_type }}" {% if case_type in  module.additional_case_types %}selected{% endif %}>{{ case_type }}</option>
+                  {% endfor %}
+                </select>
+                <input name="detail_type" type="hidden" value="{{ detail_type }}" />
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:72]: What's this? Doesn't look like it's referenced.

File: corehq/apps/app_manager/views/modules.py [hunk 4]
```diff
@@ -198,6 +198,7 @@ def _get_shared_module_view_context(request, app, module, case_property_builder,
         'case_list_form_options': _case_list_form_options(app, module, lang),
         'valid_parents_for_child_module': _get_valid_parents_for_child_module(app, module),
         'shadow_parent': _get_shadow_parent(app, module),
+        'case_types': get_case_types_for_domain(app.domain),
```
Comment (orangejenny) [inline corehq/apps/app_manager/views/modules.py:201]: Can you filter out the module's own case type?

File: corehq/apps/app_manager/xpath.py [hunk 5]
```diff
@@ -218,6 +218,16 @@ def case(self, instance_name='casedb', case_name='case'):
         quoted = CaseTypeXpath("'{}'".format(self))
         return super(CaseTypeXpath, quoted).case(instance_name, case_name)
 
+    def cases(self, additional_types=[], instance_name='casedb', case_name='case'):
+        quoted = CaseTypeXpath("'{}'".format(self))
+        selector = "{sel}={self}".format(sel=self.selector, self=quoted)
```
Comment (orangejenny) [inline corehq/apps/app_manager/xpath.py:223]: Nitpick: reusing `self` as a kwarg is a little confusing, maybe just use `quoted` as in the code below?

File: corehq/apps/app_manager/xpath.py [hunk 6]
```diff
@@ -218,6 +218,16 @@ def case(self, instance_name='casedb', case_name='case'):
         quoted = CaseTypeXpath("'{}'".format(self))
         return super(CaseTypeXpath, quoted).case(instance_name, case_name)
 
+    def cases(self, additional_types=[], instance_name='casedb', case_name='case'):
```
Comment (orangejenny) [inline corehq/apps/app_manager/xpath.py:221]: Avoid `[]` as a default arg: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: corehq/apps/case_search/utils.py [hunk 7]
```diff
@@ -99,21 +99,25 @@ def _add_daterange_queries(self):
             self.criteria.pop(key)
 
     def _add_case_property_queries(self):
-        try:
-            fuzzies = self.config.fuzzy_properties.get(
-                domain=self.domain, case_type=self.case_type).properties
-        except FuzzyProperties.DoesNotExist:
-            fuzzies = []
+        fuzzies = []
```
Comment (orangejenny) [inline corehq/apps/case_search/utils.py:102]: This is interesting: fuzzy properties are per case type, but users will be searching across case types. So if a case property is set as fuzzy for one case type but not another, and you search across both types, it'll be treated as fuzzy for both case types, right? I think that's probably fine, but definitely something to include in docs.

Comment (orangejenny) [reply corehq/apps/case_search/utils.py:102]: Same comment for the regex patterns, though I don't think anyone is using those.

File: corehq/apps/ota/views.py [hunk 8]
```diff
@@ -104,6 +104,7 @@ def app_aware_search(request, domain, app_id):
     """
     criteria = request.GET.dict()
     try:
+        # could be a list or a single string
```
Comment (orangejenny) [inline corehq/apps/ota/views.py:107]: Is this still true? It looked like you updated this to a list in a couple of places not directly to this PR, which made me think you'd switched to a list everywhere.

And if it **is** still true, would it be arduous to switch it to `case_types` and make it always a list?

**Review by orangejenny — COMMENTED** [review]
This looks great, my comments are minor.

Could you fill out the rest of the PR template and briefly summarize what the high-level final changes are? It's a little hard to follow the doc without getting bogged down in the specifics.

File: corehq/apps/ota/views.py (hunk 8 above)
Comment (sravfeyn) [reply corehq/apps/ota/views.py:107]: Yes. Any param can be a list, since we added the multiselect support. 

> would it be arduous to switch it to case_types and make it always a list?

It would require changes to the suite xml structure which always had a `case_type`. So, it's probably better to keep it as it is.

File: corehq/apps/case_search/utils.py (hunk 7 above)
Comment (sravfeyn) [reply corehq/apps/case_search/utils.py:102]: Yes, that's right. I have added this as a caveat in the spec and the product is okay with this behaviour. It's a sensible design choice, since searching across different types over same properties with same user inputs would indicate that the properties have same lexicon in most cases.

File: corehq/apps/app_manager/suite_xml/sections/remote_requests.py (hunk 1 above)
Comment (sravfeyn) [reply corehq/apps/app_manager/suite_xml/sections/remote_requests.py:140]: It would, I think since `f"{nodeset}` includes the nodeset prior, no?

File: corehq/apps/app_manager/models.py [hunk 9]
```diff
@@ -2211,6 +2211,7 @@ class ModuleBase(IndexedSchema, ModuleMediaMixin, NavMenuItemMediaMixin, Comment
     case_type = StringProperty()
     case_list_form = SchemaProperty(CaseListForm)
     module_filter = StringProperty()
+    additional_case_types = ListProperty(str)
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/app_manager/models.py:2214]: Do you mean `DetailPair`?

File: corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html (hunk 3 above)
Comment (sravfeyn) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:72]: Removed!

File: corehq/apps/app_manager/views/modules.py (hunk 4 above)
Comment (sravfeyn) [reply corehq/apps/app_manager/views/modules.py:201]: This is actually intentional. In the UI, the text for this is `Case List and Case Search Types`. This would leave the flexibility of whether to include the module's case-type in the case-list/search lists. Also it would handle the case where the user may change the module's case-type.

File: corehq/apps/app_manager/xpath.py (hunk 6 above)
Comment (sravfeyn) [reply corehq/apps/app_manager/xpath.py:221]: Done!

File: corehq/apps/app_manager/xpath.py (hunk 5 above)
Comment (sravfeyn) [reply corehq/apps/app_manager/xpath.py:223]: Sure

File: corehq/apps/app_manager/suite_xml/sections/remote_requests.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/sections/remote_requests.py:140]: aha, I misread

File: corehq/apps/app_manager/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/models.py:2214]: Yeah, `DetailPair` probably makes more sense.

[the remark this answers is not in the document]
Comment (sravfeyn) [reply corehq/apps/app_manager/models.py:2214]: I ended up putting it under `module.search_config` (`CaseSearch`)

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_case_type.html [hunk 10]
```diff
@@ -5,10 +5,17 @@
 <div class="form-group" id="case_type_form_group">
   <label class="{% css_label_class %} control-label">
     {% trans "Case Type" %}
+    {% if request|toggle_enabled:'USH_CASE_CLAIM_UPDATES' %}
+    <span class="hq-help-template"
+          data-title="{% trans "Case Type" %}"
+          data-content="{% trans "The name for the type of case you are tracking (e.g. “pregnancy,” “farm,” “household,” etc.). All cases you register in this case list will be this type of case. Other case lists with the same case type will use the same set of case(s)." %}"
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_case_type.html:11]: Minor: this should use `trans_html_attr`, which will handle quotes properly in translated text so it doesn't break the HTML.

File: corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html [hunk 11]
```diff
@@ -49,28 +49,29 @@ <h4 class="panel-title panel-title-nolink">{% trans "Menu" %}</h4>
           <div class="panel-body">
             {% if not module.is_surveys and module.module_type != 'shadow' %}
               {% include "app_manager/partials/modules/module_view_case_type.html" %}
-            {% endif %}
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:52]: Does this change affect whether or not this feature is available for shadow modules?

Comment (sravfeyn) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:52]: Yes, I changed it because I observed that the case_type setting was also under that [conditional](https://github.com/dimagi/commcare-hq/pull/29900/files#diff-d59af1ba8ae486c6decb9b89542e459d2adc916a787d971f97d2b13387b6101dR51)

Comment (orangejenny) [reply corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:52]: Got it. The case type for shadow modules gets inherited from the source module, [here](https://github.com/dimagi/commcare-hq/blob/0728e983d89ac19631a09d0bae22188e90486376/corehq/apps/app_manager/models.py#L3722-L3726), so if the additional case types can't be configured on a shadow module, they'd need to similarly inherit from the source module.

I think it's okay to implement whichever is simpler. My intuition is that shadow modules will use the same additional_case_types as the source modules but that it's not a big deal for app builders to have to specify it, especially since we have the functionality to copy an entire case list/detail config from one module to another.

**Comment by shubham1g5** [comment]
@sravfeyn should we merge these PRs for upcoming deploy ?

**Comment by shubham1g5** [comment]
@orangejenny Looks like Sravan is offline today. Do you have any cocerns merging this for the deploy tonight ?

**Comment by sravfeyn** [comment]
Yes, we should merge this today. (I am offline today, hopefully shadow modules was covered in QA)

**Comment by shubham1g5** [comment]
Looks like QA did a pass and verified that shadow module were tested. Going to merge this.
